orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+      loss_name = node.OutputDefs()[0]->Name();
+      return;
+    }
+  }
+}
+
+bool IsBackward(Node& node) {
+  return (node.Description() == "Backward pass");
+}
+
+void AddInputEvent(Graph& graph, const std::string& op_name,
+                   bool is_forward,
+                   std::vector<NodeArg*>& input_args,
+                   std::vector<std::string>& new_input_names) {
+  ONNX_NAMESPACE::TypeProto event_type_proto;
+  event_type_proto.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+
+  auto event_id_name = graph.GenerateNodeArgName(op_name + (is_forward ? "_fw" : "_bw") + "_event_id");
+  auto& event_id = graph.GetOrCreateNodeArg(event_id_name, &event_type_proto);
+  new_input_names.push_back(event_id_name);
+
+  input_args.push_back(&event_id);
+}
+
+void AddRecordBackward(Graph& graph, Node* send_bw, std::vector<std::string>& new_input_names) {
+  // gradient graph can contain some dangling leaf nodes. Add them all to WaitEvent
+  // backward node's input.
+  auto add_leaf_node = [&](std::vector<NodeArg*>& input_args) {
+    for (auto& node : graph.Nodes()) {
+      if (!IsBackward(node)) {
+        // only check backward node
+        continue;
+      }
+      bool find_consumer_nodes = false;
+      std::vector<NodeArg*>& outputs = node.MutableOutputDefs();
+      for (auto& output : outputs) {
+        std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(output->Name());
+        if (consumer_nodes.size() > 0) {
+          find_consumer_nodes = true;
+          break;
+        }
+      }
+      if (!find_consumer_nodes && outputs.size() > 0) {
+        input_args.push_back(outputs[0]);
+      }
+    }
+  };
+
+  std::vector<NodeArg*> input_args;
+  AddInputEvent(graph, "RecordEvent", false /* is_forward */, input_args, new_input_names);
+
+  if (send_bw) {
+    // if we have send op in backward pass (at the end of the graph), add the RecordEvent op after that.
+    input_args.insert(std::end(input_args),
+                      std::begin(send_bw->MutableOutputDefs()),
+                      std::end(send_bw->MutableOutputDefs()));
+  }
+  add_leaf_node(input_args);
+  graph.AddNode(graph.GenerateNodeName("RecordEvent"),
+                "RecordEvent",
+                "Backward pass",
+                input_args,
+                {},
+                nullptr,
+                kMSDomain);
+}
+
+Status AddWaitForward(Graph& graph, Node* recv_fw, std::vector<std::string>& new_input_names) {
+  auto update_wait_input_output = [&](NodeArg* old_input,
+                                      std::vector<NodeArg*>& input_args,
+                                      std::vector<NodeArg*>& output_args) -> NodeArg& {
+    const auto& new_name = graph.GenerateNodeArgName(old_input->Name());
+    ONNX_NAMESPACE::TypeProto type_proto(*(old_input->TypeAsProto()));
+
+    auto& wait_output = graph.GetOrCreateNodeArg(new_name, &type_proto);
+    output_args.push_back(&wait_output);
```

Comment:
`push_back` a pointer of local variable?

Reply:
wait_output is not a local variable. It is a reference of a node that is created and added into graph. So the lifetime of wait_output is controled by graph object and it won't get distroyed after existing this function. 

---
In reply to: [409776283](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409776283) [](ancestors = 409776283)